lightning_examples/warp-drive/multi_agent_rl.py

```diff
@@ -0,0 +1,343 @@
+    warp_drive_module,
+    fps=25,
+    tagger_color="#C843C3",
+    runner_color="#245EB6",
+    runner_not_in_game_color="#666666",
+    fig_width=6,
+    fig_height=6,
+):
+    assert warp_drive_module is not None
+    episode_states = warp_drive_module.fetch_episode_states(["loc_x", "loc_y", "still_in_the_game"])
+    assert isinstance(episode_states, dict)
+    env = warp_drive_module.cuda_envs.env
+
+    fig, ax = plt.subplots(1, 1, figsize=(fig_width, fig_height))  # , constrained_layout=True
+    ax.remove()
+    ax = fig.add_subplot(1, 1, 1, projection="3d")
+
+    # Bounds
+    ax.set_xlim(0, 1)
+    ax.set_ylim(0, 1)
+    ax.set_zlim(-0.01, 0.01)
+
+    # Surface
+    corner_points = [
+        (0, 0),
+        (0, 1),
+        (1, 1),
+        (1, 0),
+    ]
+
+    poly = Polygon(corner_points, color=(0.1, 0.2, 0.5, 0.15))
+    ax.add_patch(poly)
+    art3d.pathpatch_2d_to_3d(poly, z=0, zdir="z")
+
+    # "Hide" side panes
+    ax.xaxis.set_pane_color((1.0, 1.0, 1.0, 0.0))
+    ax.yaxis.set_pane_color((1.0, 1.0, 1.0, 0.0))
+    ax.zaxis.set_pane_color((1.0, 1.0, 1.0, 0.0))
+
+    # Hide grid lines
+    ax.grid(False)
+
+    # Hide axes ticks
+    ax.set_xticks([])
+    ax.set_yticks([])
+    ax.set_zticks([])
+
+    # Hide axes
+    ax.set_axis_off()
+
+    # Set camera
+    ax.elev = 40
+    ax.azim = -55
+    ax.dist = 10
+
+    # Try to reduce whitespace
+    fig.subplots_adjust(left=0, right=1, bottom=-0.2, top=1)
+
+    # Plot init data
+    lines = [None for _ in range(env.num_agents)]
+
+    for idx in range(len(lines)):
+        if idx in env.taggers:
+            lines[idx] = ax.plot3D(
+                episode_states["loc_x"][:1, idx] / env.grid_length,
+                episode_states["loc_y"][:1, idx] / env.grid_length,
+                0,
+                color=tagger_color,
+                marker="o",
+                markersize=10,
+            )[0]
+        else:  # runners
+            lines[idx] = ax.plot3D(
+                episode_states["loc_x"][:1, idx] / env.grid_length,
+                episode_states["loc_y"][:1, idx] / env.grid_length,
+                [0],
+                color=runner_color,
+                marker="o",
+                markersize=5,
+            )[0]
+
+    init_num_runners = env.num_agents - env.num_taggers
+
+    def _get_label(timestep, n_runners_alive, init_n_runners):
+        line1 = "Continuous Tag\n"
+        line2 = "Time Step:".ljust(14) + f"{timestep:4.0f}\n"
+        frac_runners_alive = n_runners_alive / init_n_runners
+        pct_runners_alive = f"{n_runners_alive:4} ({frac_runners_alive * 100:.0f}%)"
+        line3 = "Runners Left:".ljust(14) + pct_runners_alive
+        return line1 + line2 + line3
+
+    label = ax.text(
+        0,
+        0,
+        0.02,
+        _get_label(0, init_num_runners, init_num_runners).lower(),
+    )
+
+    label.set_fontsize(14)
+    label.set_fontweight("normal")
+    label.set_color("#666666")
+
+    def animate(i):
```

Comment:
consider move these function out of this parent func so it is not so heavy

Reply:
Will leave it as is, and do no wish to expose this outside.